configuration/config2/EnvironmentEvents.cpp

```diff
@@ -0,0 +1,106 @@
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+############################################################################## */
+
+#include "EnvironmentEvents.hpp"
+#include "EnvironmentNode.hpp"
+#include "EnvironmentValue.hpp"
+
+void AttributeDependencyCreateEvent::addDependency(const std::string &attrName, const std::string &attrVal, const std::string &depAttr, const std::string &depVal)
+{
+    auto valSetMapIt = m_depAttrVals.find(attrName);
+    if (valSetMapIt == m_depAttrVals.end())
+    {
+        std::map<std::string, std::pair<std::string, std::string>> keyValMap;
+        m_depAttrVals[attrName] = keyValMap;
+    }
+
+    m_depAttrVals[attrName][attrVal] = std::pair<std::string, std::string>(depAttr, depVal);
+}
+
+
+bool CreateEnvironmentEvent::handleEvent(const std::string &eventType, std::shared_ptr<EnvironmentNode> pEventNode)
+{
+    return pEventNode->getSchemaItem()->getItemType() == m_itemType;
+}
+
+
+bool AttributeDependencyCreateEvent::handleEvent(const std::string &eventType, std::shared_ptr<EnvironmentNode> pEventNode)
+{
+    bool rc = false;
+    if (CreateEnvironmentEvent::handleEvent(eventType, pEventNode))
+    {
+        for (auto attrIt = m_depAttrVals.begin(); attrIt != m_depAttrVals.end(); ++attrIt)
+        {
+            std::shared_ptr<EnvironmentValue> pAttr = pEventNode->getAttribute(attrIt->first);
+            if (pAttr && pAttr->getSchemaValue()->getType()->isEnumerated())
+            {
+                rc = true;   // we handled at least one
+                for (auto valueIt = attrIt->second.begin(); valueIt != attrIt->second.end(); ++valueIt)
+                {
+                    pAttr->getSchemaValue()->getType()->getLimits()->addDependentAttributeValue(valueIt->first, valueIt->second.first, valueIt->second.second);
+                }
+            }
+        }
+    }
+    return rc;
+}
+
+
+bool InsertEnvironmentDataCreateEvent::handleEvent(const std::string &eventType, std::shared_ptr<EnvironmentNode> pEventNode)
+{
+    bool rc = false;
+    if (CreateEnvironmentEvent::handleEvent(eventType, pEventNode))
+    {
+        if (!m_itemAttribute.empty())
+        {
+            std::vector<std::shared_ptr<EnvironmentNode>> matchNodes;
+
+            if (!m_matchPath.empty())
+            {
+                pEventNode->findNodes(m_matchPath, matchNodes);
+            }
+            else
+            {
+                matchNodes.push_back(pEventNode);
+            }
+
+            for (auto nodeIt = matchNodes.begin(); nodeIt != matchNodes.end(); ++nodeIt)
+            {
+                if (!m_itemAttribute.empty())
+                {
+                    std::shared_ptr<EnvironmentValue> pItemAttr = pEventNode->getAttribute(m_itemAttribute);
```

Comment:
Should cases like these be "unique_ptr" rather than shared_ptr?  If so, there are probably many cases throughout.

Reply:
In most cases the called method returns a shared_ptr since because the object already exists and most likely has a reference. Converting to a unique_ptr in these cases would not follow the architecture since unique ownership is not the intent. I will, however, keep an eye out for situations where a unique_ptr would suffice, such as where the method creates the object and persistence is not necessary.